### PR TITLE
Add a search-and-replace diff strategy

### PR DESCRIPTION
**Pull request by mrubens**
This is a new strategy for performing diff edits, similar to what Aider and Codey use. It seems to work a lot better than unified diffs so far 🙏 

----

> [!IMPORTANT]
> Introduces `SearchReplaceDiffStrategy` for precise search and replace operations, replacing `UnifiedDiffStrategy`, with updates to documentation and tests.
> 
>   - **Behavior**:
>     - Introduces `SearchReplaceDiffStrategy` in `search-replace.ts` for precise search and replace operations.
>     - Updates `getDiffStrategy()` in `DiffStrategy.ts` to use `SearchReplaceDiffStrategy` for all models.
>   - **Testing**:
>     - Adds `search-replace.test.ts` to test `applyDiff` and `getToolDescription` methods of `SearchReplaceDiffStrategy`.
>     - Tests handle cases like matching content, whitespace handling, invalid formats, and directory inclusion.
>   - **Documentation**:
>     - Updates `CHANGELOG.md` to version 2.1.17 with new diff strategy details.
>     - Updates `README.md` to include support for copying prompts from the history screen.
>     - Bumps version in `package.json` to 2.1.17.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=RooVetGit%2FRoo-Cline&utm_source=github&utm_medium=referral)<sup> for 205bb5484c1cebd5edb7204143ca142c4e3d45bf. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: src/core/diff/strategies/search-replace.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        // Trim the original content for comparison
+        const trimmedOriginal = originalContent.trim();
+        
+        // Verify the search content exists in the trimmed original
+        if (!trimmedOriginal.includes(trimmedSearch)) {
+            return false;
+        }
+
+        // Replace the content, maintaining original whitespace
+        return originalContent.replace(trimmedSearch, trimmedReplace);
```
[the remark this answers is not in the document]
Comment (mrubens) [reply src/core/diff/strategies/search-replace.ts:86]: I think it's safer to just do the first one

File: src/core/diff/strategies/search-replace.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        }
+
+        const [_, searchContent, replaceContent] = match;
+        
+        // Trim both search and replace content
+        const trimmedSearch = searchContent.trim();
+        const trimmedReplace = replaceContent.trim();
+        
+        // Trim the original content for comparison
+        const trimmedOriginal = originalContent.trim();
```
[the remark this answers is not in the document]
Comment (mrubens) [reply src/core/diff/strategies/search-replace.ts:78]: This is intentional - the indentation can be slightly off sometimes

[the remark this answers is not in the document]
Comment (mrubens) [reply src/core/diff/strategies/search-replace.ts:78]: This is intentional

File: src/core/diff/strategies/search-replace.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    applyDiff(originalContent: string, diffContent: string): string | false {
+        // Extract the search and replace blocks
+        const match = diffContent.match(/<<<<<<< SEARCH\n([\s\S]*?)\n=======\n([\s\S]*?)\n>>>>>>> REPLACE/);
+        if (!match) {
+            return false;
+        }
+
+        const [_, searchContent, replaceContent] = match;
+        
+        // Trim both search and replace content
```
Comment (mrubens) [inline src/core/diff/strategies/search-replace.ts:73]: Instead of trim should probably collapse all white space to match multi line better

**Review by ColemanRoo — APPROVED** [review]

**Pull request merged**